runtime/v2/shim_load.go

```diff
@@ -88,18 +88,35 @@ func (m *ShimManager) loadShims(ctx context.Context) error {
 			bundle.Delete()
 			continue
 		}
-		container, err := m.containers.Get(ctx, id)
-		if err != nil {
-			log.G(ctx).WithError(err).Errorf("loading container %s", id)
-			if err := mount.UnmountAll(filepath.Join(bundle.Path, "rootfs"), 0); err != nil {
-				log.G(ctx).WithError(err).Errorf("forceful unmount of rootfs %s", id)
+
+		var (
+			runtime string
+		)
+
+		// If we're on 1.6+ and specified custom path to the runtime binary, path will be saved in 'runtime' file.
+		if data, err := os.ReadFile(filepath.Join(bundle.Path, "runtime")); err == nil {
+			runtime = string(data)
+		} else if err != nil && !os.IsNotExist(err) {
+			log.G(ctx).WithError(err).Error("failed to read `runtime` path from bundle")
+		}
+
+		// Query runtime name from metadata store
+		if runtime == "" {
+			container, err := m.containers.Get(ctx, id)
```

Comment:
Just find one leaky resource case can be solved by this change. 

There is the case: If the shim is killed somehow, the container.Delete will ignore that not-found error and delete the data. And then containerd has been restarted and just deletes the bundle without calling shim cleanup. 

In this change, the containerd will try to connect to the shim and find that is dead and call cleanup. 😀